src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -14,13 +18,62 @@
   "v3")
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Private
+
+(defn class-info-from-jruby->class-info-for-json
+  "Convert a class info map received from the jruby service into an
+  appropriate map for use in serializing an environment_classes response to
+  JSON"
+  [info-from-jruby environment]
+  (->> info-from-jruby
+       (map #(hash-map :path (key %) :classes (val %)))
+       (sort-by :path)
+       (into [])
+       (hash-map :name environment :files)))
+
+(defn get-environment-class-info
+  "Middleware function for constructing a Ring response from an incoming
+  request for environment_classes information"
+  [jruby-service]
+  (fn [request]
+    (if-let [environment (jruby-request/get-environment-from-request request)]
```

Comment:
Yeah, I did think about just having the `environment` be a subpath under the `environment_classes` endpoint.  I agree that would look better.  But I thought that for "consistency" it would be better to just have the environment in a query parameter for now.  It felt really confusing to me to be curling different endpoints under /puppet/v3 and having to remember that the `environment_classes` does it as part of the URL whereas all of the others use a query parameter.

Maybe /puppet/v4 would provide a good opportunity to restructure all of the environment-aware URLs to incorporate the environment into the URL subpath?